File: lib/papagaio_tts.dart
import 'dart:ui';

import 'papagaio_tts_platform_interface.dart';

class PapagaioTts {
  Future<List<String>> getVoices() {
    var result = PapagaioTtsPlatform.instance.getVoices();
    return Future<List<String>>.value(result);
  }

  Future<List<Locale>> getAvailableLanguages(List<String>? filterLanguages) async {
    List<dynamic> result = await PapagaioTtsPlatform.instance.getAvailableLanguages(filterLanguages);
    var localeList = result.map((objList) {
      return Locale(objList[0], objList[1]);
    }).toList();
    return Future<List<Locale>>.value(localeList);
  }

  Future<bool> speak(String text) {
    final result = PapagaioTtsPlatform.instance.speak(text);
    return Future<bool>.value(result);
  }

  Future<bool> stop() {
    final result = PapagaioTtsPlatform.instance.stop();
    return Future<bool>.value(result);
  }

  Future<void> shutdown() {
    PapagaioTtsPlatform.instance.shutdown();
    return Future<void>.value();
  }

  Future<bool> getSpeakingStatus() {
    return PapagaioTtsPlatform.instance.getSpeakingStatus();
  }

  Future<Locale> getLanguage() async {
    String result = await PapagaioTtsPlatform.instance.getLanguage();
    List<String> langCountry = result.split(RegExp("-|_"));
    return Future<Locale>.value(Locale(langCountry[0], langCountry[1]));
  }

  Future<String> getVoice() {
    return PapagaioTtsPlatform.instance.getVoice();
  }

  Future<num> getRate() {
    return PapagaioTtsPlatform.instance.getRate();
  }

  Future<num> getVolume() {
    return PapagaioTtsPlatform.instance.getVolume();
  }

  Future<num> getPitch() {
    return PapagaioTtsPlatform.instance.getPitch();
  }

  Future<bool> setVoice(String voiceName) {
    final result = PapagaioTtsPlatform.instance.setVoice(voiceName);
    return Future<bool>.value(result);
  }

  Future<bool> setLanguage(Locale language) {
    final result = PapagaioTtsPlatform.instance.setLanguage("${language.languageCode}-${language.countryCode}");
    return Future<bool>.value(result);
  }

  Future<bool> setRate(num rate) {
    final result = PapagaioTtsPlatform.instance.setRate(rate);
    return Future<bool>.value(result);
  }

  Future<bool> setVolume(num volume) {
    final result = PapagaioTtsPlatform.instance.setVolume(volume);
    return Future<bool>.value(result);
  }

  Future<bool> setPitch(num pitch) {
    final result = PapagaioTtsPlatform.instance.setPitch(pitch);
    return Future<bool>.value(result);
  }
}

File: android/src/main/kotlin/papagaiojp/webnode/jp/papagaio_tts/PapagaioTts.kt
package papagaiojp.webnode.jp.papagaio_tts


import android.content.Context
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.speech.tts.Voice
import java.util.Locale


class PapagaioTts(context: Context) {
    private lateinit var textToSpeech: TextToSpeech

    init {
        textToSpeech = TextToSpeech(context) { status ->
            if (status == TextToSpeech.SUCCESS) {
                textToSpeech.language = Locale.ENGLISH
            } else {
                println("ERROR: status = $status")
                throw Exception("Failure with initialize")
            }
        }
    }

    fun speak(text: String, volume: Float): Boolean {
        val params = Bundle()
        params.putFloat(TextToSpeech.Engine.KEY_PARAM_VOLUME, volume)
        val result = textToSpeech.speak(text, TextToSpeech.QUEUE_FLUSH, params, "")
        return result == TextToSpeech.SUCCESS
    }

    fun getAvailableLanguages(filterLanguages: List<String>): List<List<String>> {
        val allLanguages = textToSpeech.availableLanguages.map {it.toString().split("_")}.toList()
        if (filterLanguages.isEmpty()) {
            return allLanguages
        }

        return allLanguages.filter { filterLanguages.contains( it[0] ) }.toList()
    }

    fun getVoices(): List<String> {
        val lang: String = getLanguage()
        val voices: Set<Voice> = textToSpeech.voices

        return voices.filter { it.locale.toString().startsWith(lang) }.map { it.name }
            .toList()
    }

    fun getVoice(): String {
        val currentVoice = textToSpeech.voice
        if (currentVoice != null) {
            return currentVoice.name
        }
        return textToSpeech.defaultVoice.name ?: ""
    }

    fun getLanguage(): String {
        if(textToSpeech.voice != null) {
            return textToSpeech.voice.locale.toString()
        }
        return Locale.ENGLISH.toString()
    }

    fun getSpeakingStatus(): Boolean {
        return textToSpeech.isSpeaking
    }

    fun setVoice(voiceName: String): Boolean {
        val voice: Voice = textToSpeech.voices.filter { it.name == voiceName }[0]
        val result = textToSpeech.setVoice(voice)
        return result == TextToSpeech.SUCCESS
    }

    fun setLanguage(language: String, country: String): Boolean {
        val locale = Locale(language, country)
        val langLocale = Locale(language)
        if (textToSpeech.isLanguageAvailable(locale) == TextToSpeech.LANG_AVAILABLE) {
            val result = textToSpeech.setLanguage(locale)
            return result == TextToSpeech.SUCCESS
        } else if(textToSpeech.isLanguageAvailable(langLocale) == TextToSpeech.LANG_AVAILABLE) {
            val result = textToSpeech.setLanguage(langLocale)
            return result == TextToSpeech.SUCCESS
        }
        return false
    }

    fun setPitch(pitch: Float): Boolean {
        val result = textToSpeech.setPitch(pitch)
        return result == TextToSpeech.SUCCESS
    }

    fun setRate(rate: Float): Boolean {
        // Rate for Swift has limit 1.0 and Android 2.0
        val result = textToSpeech.setSpeechRate(rate * 2)
        return result == TextToSpeech.SUCCESS
    }
    fun stop(): Boolean {
        val result = textToSpeech.stop()
        return result == TextToSpeech.SUCCESS
    }

    fun shutdown() {
       textToSpeech.shutdown()
    }
}